### PR TITLE
[SCB-548] support gracefully shutdown

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
When user call system.exit(0),these three works will do:
1.Unregister microservice instance from Service Center:
this work will do by process ContextClosedEvent, we need unregister immediately for stopping provide service any more.
2.Waiting for all invocations to finish:
this work will do by ShutdownHandler, when all invocations finished or 'deadline time reached', cc and transport vertx threads will close.
3.All spring bean do close process
we had registerShutdownHook for spring ApplicationContext and beans can define 'destroy-method' do cleaning.

here is example:
 
Spring mvc Hello Java Chassis
Pojo Hello person ServiceComb/Java Chassis
Jaxrs Hello person ServiceComb/Java Chassis
Spring mvc Hello person ServiceComb/Java Chassis
2018-05-09 16:02:14,288 [WARN] handler chain is shutting down org.apache.servicecomb.core.handler.ShutdownHookHandler.run(ShutdownHookHandler.java:87)
2018-05-09 16:02:14,289 [INFO] Closing org.springframework.context.support.ClassPathXmlApplicationContext@7cf10a6f: startup date [Wed May 09 16:02:05 CST 2018]; root of context hierarchy org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:984)
2018-05-09 16:02:14,290 [WARN] cse is closing now... org.apache.servicecomb.core.CseApplicationListener.onApplicationEvent(CseApplicationListener.java:148)
2018-05-09 16:02:14,291 [INFO] service center task is shutdown. org.apache.servicecomb.serviceregistry.registry.RemoteServiceRegistry.onShutdown(RemoteServiceRegistry.java:72)
2018-05-09 16:02:14,295 [WARN] handler chain is shut down org.apache.servicecomb.core.handler.ShutdownHookHandler.run(ShutdownHookHandler.java:103)
2018-05-09 16:02:14,296 [INFO] Unregister microservice instance success. microserviceId=90b76fd551c511e8b51db4b676a39f40 instanceId=481f630d535f11e8bc19b4b676a39f40 org.apache.servicecomb.serviceregistry.registry.AbstractServiceRegistry.unregisterInstance(AbstractServiceRegistry.java:232)
Process finished with exit code 0

### CODE REVIEW

**Comment by zhengyangyong** [comment]
Springmvc it test failed because we had fully shutdown and re-init seems do not init transport again,I will try fix this problem

**Comment by WillemJiang** [comment]
As the Server socket maybe reused, it's hard to reconnect the server which is restarted with the same port. I think we need to avoid restarting the server in the unit test or system test.

File: core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java [hunk 1]
```diff
@@ -56,6 +57,8 @@ public long getActiveCount() {
   @Override
   public void handle(Invocation invocation, AsyncResponse asyncResp) throws Exception {
     if (shuttingDown) {
+      System.out.println("shutting down in progress");
```
Comment (liubao68) [inline core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java]: Do not use System.out.println to print logs. And we have already throw an exception with this message, and I think this log is not necessary.

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java]: So sorry is for debug,thanks!

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java]: Deleted

**Comment by zhengyangyong** [comment]
[[1;31mERROR[m] Failed to execute goal [32mio.fabric8:docker-maven-plugin:0.20.0:start[m [1m(start)[m on project [36mdynamic-config-tests[m: [1;31mExecution start of goal io.fabric8:docker-maven-plugin:0.20.0:start failed: Start-Job failed with unexpected exception: [nobodyiam/apollo-quick-start] "apollo.servicecomb.apache.org": Timeout after 120033 ms while waiting on log out 'Portal started' and on tcp port '[/172.17.0.3:8080, /172.17.0.3:8070]'[m -> [1m[Help 1][m

Failed by apollo IT

**Comment by wujimin** [comment]
if ContextCloseEvent and handler/ShutdownHookHandler.java is random ordered, then it's not ok

**Comment by zhengyangyong** [comment]
I had used a Semaphore in order to sync all invocation had finished.now gracefully shutdown had fixed below :

1. Unregister microservice instance from Service Center and close vertx
2. Wait all invocation finished (delete timeout mechanism in ShutdownHandler)
3. Notify all component do clean works via Event
4. Stop vertx to prevent blocking exit

**Comment by zhengyangyong** [comment]
Rebase on latest master

**Comment by zhengyangyong** [comment]
improve gracefully shutdown process order
```java
  private void gracefullyShutdown() {
    //Step 1: notify all component stop invoke via BEFORE_CLOSE Event
    triggerEvent(EventType.BEFORE_CLOSE);

    //Step 2: Unregister microservice instance from Service Center and close vertx
    RegistryUtils.destroy();
    VertxUtils.closeVertxByName("registry");

    //Step 3: wait all invocation finished
    try {
      ShutdownHookHandler.INSTANCE.ALL_INVOCATION_FINISHED.acquire();
      LOGGER.warn("all invocation finished");
    } catch (InterruptedException e) {
      LOGGER.error("invocation finished semaphore interrupted", e);
    }

    //Step 4: Stop vertx to prevent blocking exit
    VertxUtils.closeVertxByName("config-center");
    VertxUtils.closeVertxByName("transport");

    //Step 5: notify all component do clean works via AFTER_CLOSE Event
    triggerEvent(EventType.AFTER_CLOSE);
  }
```

**Comment by wujimin** [comment]
ShutdownHookHandler is not so good, sometimes will block shutdown process cause of wrong statistics, must depend on timeout to make shutdown process continue.

we should:
1.prevent create new consumer invocation
2.prevent other find me
3.make accurate statistics

**Comment by wujimin** [comment]
and CI failed, it's about trace, i can find what happened, it seems that did not print error test case name?

**Comment by zhengyangyong** [comment]
@wujimin failed by 'ExecutionException Error occurred in starting fork, check output in log', and I do not change zuul test case, I will reopen for restart ci check

File: core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java [hunk 2]
```diff
@@ -39,13 +40,16 @@
...
-
-  private final int period = 10;
-
   private volatile boolean shuttingDown = false;
 
+  public final Semaphore ALL_INVOCATION_FINISHED = new Semaphore(1);
+
   private ShutdownHookHandler() {
+    try {
+      ALL_INVOCATION_FINISHED.acquire();
```
Comment (liubao68) [inline core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java]: I think it's better to make semaphore private and add a method like waitInvocationFinish.  And in initialization, acquire is not required and shutdownhook can be removed too.

File: core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java [hunk 3]
```diff
@@ -62,36 +66,34 @@ public void handle(Invocation invocation, AsyncResponse asyncResp) throws Except
...
         }
       });
     } catch (Throwable e) {
       responseCounter.incrementAndGet();
+      validAllInvocationFinished();
       throw e;
     }
   }
 
+  private synchronized void validAllInvocationFinished() {
```
Comment (liubao68) [inline core/src/main/java/org/apache/servicecomb/core/handler/ShutdownHookHandler.java]: This synchronized code blocking all invocations and not good idea

**Comment by liubao68** [comment]
@wujimin 

ShutdownHookHandler is not so good, sometimes will block shutdown process cause of wrong statistics, must depend on timeout to make shutdown process continue. ----I don't know how this comes. Because 
ShutdownHookHandler will wait quit a long time much more than the request timeout. If the time reached there usually means timeout not properly set or some other bug. 

we should:
1.prevent create new consumer invocation  ----This is have already done
2.prevent other find me  ---We can not prevent it from actually happen and maybe time delay.
3.make accurate statistics ---The process is shutdown and we force close some unexpected invocations and I think the precision is not very important and accurate statistics can not be achieved.

We are experiencing a lot of scenarios that can not be shutdown but need force to exit the program. Program not exit can cause more bad consequences.

**Comment by wujimin** [comment]
in performance test, ShutdownHookHandler will always block shutdown process for many seconds (wait for timeout for wrong request/response count statistics)

will change statistics from invocation handler to invocation event
after change to invocation event, then reject in invocation handler is too late
so shutdown hook will rewrite

by new hook, even in performance test, we can shutdown normally in one or two second.

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  @Subscribe
+  public void onInvocationFinish(InvocationFinishEvent event) {
+    invocationFinishedCounter.incrementAndGet();
+    if (validIsStopping()) {
+      validAllInvocationFinished();
+    }
+  }
+
+  public synchronized void init() {
+    if (validIsDown()) {
```
Comment (WillemJiang) [inline core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: It's better to use the SCBStatus directly , I cannot get the method work by reading the name.

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: sounds good , fixed

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    //Step 4: Stop vertx to prevent blocking exit
+    VertxUtils.blockCloseVertxByName("config-center");
+    VertxUtils.blockCloseVertxByName("transport");
+
+    //Step 5: notify all component do clean works via AFTER_CLOSE Event
+    safeTriggerEvent(EventType.AFTER_CLOSE);
+
+    //Step 6: Clean flags for re-init
+    eventBus.unregister(this);
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: when succeed to init, then already unregister this, unregister a not register subscribe will cause exception.

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: yes,it had unregister in triggerAfterRegistryEvent after succeed to init,I will delete this line

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        doUninit();
+        status = SCBStatus.DOWN;
+      } catch (Exception e) {
+        status = SCBStatus.FAILED;
+        LOGGER.info("ServiceComb failed shutdown", e);
+      }
+    }
+  }
+
+  private void doUninit() throws Exception {
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: i think this should not throw exception
even one step failed, must finish remain steps

File: core/src/main/java/org/apache/servicecomb/core/SCBStatus.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  //Chassis is Down
+  DOWN,
+  //Chassis is Starting (progressing)
+  STARTING,
+  //Chassis is Running
+  UP,
+  //Chassis is Stopping (progressing)
+  STOPPING,
+  //Chassis Init Failed
+  FAILED
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/SCBStatus.java]: seems that uninit failed, will set status to FAILED.
so what's your design?

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBStatus.java]: I don't know if uninit failed, can enable init again ,so I think we may better set to FAILED ?

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java (hunk 6 above)
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: OK, i will fix

File: core/src/main/java/org/apache/servicecomb/core/SCBStatus.java (hunk 7 above)
Comment (wujimin) [reply core/src/main/java/org/apache/servicecomb/core/SCBStatus.java]: currently,  we will only init and uninit only one time
but i remember your UT will do this multi times?

**Comment by zhengyangyong** [comment]
Failed by apollo it , [[1;31mERROR[m] Failed to execute goal [32mio.fabric8:docker-maven-plugin:0.20.0:start[m [1m(start)[m on project [36mdynamic-config-tests[m: [1;31mExecution start of goal io.fabric8:docker-maven-plugin:0.20.0:start failed: Start-Job failed with unexpected exception: [nobodyiam/apollo-quick-start] "apollo.servicecomb.apache.org": Timeout after 120162 ms while waiting on log out 'Portal started' and on tcp port '[/172.17.0.3:8080, /172.17.0.3:8070]'[m -> [1m[Help 1][m Reopen

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+    status = SCBStatus.STOPPING;
+
+    //Step 2: Unregister microservice instance from Service Center and close vertx
+    // Forbidden other consumers find me
+    RegistryUtils.destroy();
+    VertxUtils.blockCloseVertxByName("registry");
+
+    //Step 3: wait all invocation finished
+    // forbit create new consumer invocation
```
Comment (wujimin) [inline core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: forbit comments should move to  status = SCBStatus.STOPPING;

File: core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java [hunk 9]
```diff
@@ -32,31 +35,31 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(InvokerUtils.class);
 
   public static Object syncInvoke(String microserviceName, String schemaId, String operationName, Object[] args) {
-    ReferenceConfig referenceConfig = ReferenceConfigUtils.getForInvoke(microserviceName);
-    SchemaMeta schemaMeta = referenceConfig.getMicroserviceMeta().ensureFindSchemaMeta(schemaId);
-    Invocation invocation = InvocationFactory.forConsumer(referenceConfig, schemaMeta, operationName, args);
-    return syncInvoke(invocation);
+    validCanInvoke();
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java]: Considering the check cost is low, I think it's no problem becasuse these method are all public and user may direct use them for general invoke.

[the remark this answers is not in the document]
Comment (wujimin) [reply core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java]: ok,will changed in future PR

File: core/src/main/java/org/apache/servicecomb/core/SCBEngine.java (hunk 8 above)
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: yes,it's my mistake,will fix

Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/SCBEngine.java]: Done

File: core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (WillemJiang) [reply core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java]: "validCanInvoke" is a bad name (It just check the status of Engine), how about using the "checkEngineStatus".

[the remark this answers is not in the document]
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/provider/consumer/InvokerUtils.java]: Done

**Pull request merged**